Content.Server/_NC/Trade/Store/Runtime/Transactions/NcStoreLogicSystem.Transactions.cs

```diff
@@ -0,0 +1,220 @@
+            return false;
+        }
+
+        var deliveredPurchases = spawnedUnits / unitsPerPurchase;
+        if (deliveredPurchases <= 0)
+        {
+            GiveCurrency(user, currency, totalPrice);
+            return false;
+        }
+
+        if (deliveredPurchases < purchases)
+        {
+            var refundPurchases = purchases - deliveredPurchases;
+            var refundL = (long) refundPurchases * unitPrice;
+            if (refundL > 0 && refundL <= int.MaxValue)
+                GiveCurrency(user, currency, (int) refundL);
+        }
+
+        if (listing.RemainingCount >= 0)
+            listing.RemainingCount = Math.Max(0, listing.RemainingCount - deliveredPurchases);
+
+        Sawmill.Info(
+            $"TryBuy: OK {listing.ProductEntity} x{spawnedUnits} ({deliveredPurchases} purchases) for {unitPrice} {currency} each");
+        return true;
+
+    }
+
+    public bool TrySell(string listingId, EntityUid machine, NcStoreComponent? store, EntityUid user, int count = 1)
+    {
+        if (store == null)
+            return false;
+        return TrySellScenario(listingId, store, user, user, count, out _);
+    }
+
+    public bool TrySellFromContainer(
+        string listingId,
+        EntityUid machine,
+        NcStoreComponent? store,
+        EntityUid user,
+        EntityUid container,
+        int count = 1
+    )
+    {
+        if (store == null)
+            return false;
+        return TrySellScenario(listingId, store, user, container, count, out var sold) &&
+            LogSellFromContainer(sold, listingId, store, container);
+    }
+
+    private bool TrySellScenario(
+        string listingId,
+        NcStoreComponent store,
+        EntityUid user,
+        EntityUid root,
+        int count,
+        out int sold
+    )
+    {
+        sold = 0;
+        if (store.Listings.Count == 0 || count <= 0)
+            return false;
+        if (!store.ListingIndex.TryGetValue(
+            NcStoreComponent.MakeListingKey(StoreMode.Sell, listingId),
+            out var listing))
+            return false;
+        if (!TryPickCurrencyForSell(store, listing, out var currency, out var unitPrice) || unitPrice <= 0)
+            return false;
+
+        _inventory.InvalidateInventoryCache(root);
+
+        // Считаем через новую систему
+        var snap = _inventory.BuildInventorySnapshot(root);
+        var owned = _inventory.GetOwnedFromSnapshot(snap, listing.ProductEntity, listing.MatchMode);
+
+        var maxByRemaining = listing.RemainingCount >= 0 ? listing.RemainingCount : int.MaxValue;
+        var maxPossible = Math.Min(owned, maxByRemaining);
+        if (maxPossible <= 0)
+            return false;
+
+        var actual = Math.Min(count, maxPossible);
+
+        // Забираем через новую систему
+        var ok = _inventory.TryTakeProductUnitsFromRootCached(root, listing.ProductEntity, actual, listing.MatchMode);
+        if (!ok)
+            return false;
```

Comment:
```suggestion
        if (!_inventory.TryTakeProductUnitsFromRootCached(root, listing.ProductEntity, actual, listing.MatchMode))
            return false;
```
а зачем вводить переменную которую потом нигде не используешь кроме иф проверки